plugins/woocommerce/src/Utilities/RestApiUtil.php

```diff
@@ -27,4 +27,89 @@ public function get_endpoint_data( $endpoint, $params = array() ) {
+		 * @param string $route_namespace            The namespace.
+		 *
+		 * @since 10.3.0
+		 */
+		$should_lazy_load_namespace = apply_filters( 'woocommerce_rest_should_lazy_load_namespace', true, $route_namespace );
+		if ( $should_lazy_load_namespace ) {
+			$this->attach_lazy_loaded_namespace( $route_namespace, $callback );
+		} else {
+			call_user_func( $callback );
+		}
+	}
+
+	/**
+	 * This is the internal function that implements the logic of wc_rest_lazy_load_namespace(). Its interface
+	 * and behavior is not guaranteed.  It solely exists so that $callback_filter_id does not need to be part of the
+	 * public interface to `wc_rest_lazy_load_namespace()`. Do not call it directly.
+	 *
+	 * @param string   $route_namespace    The namespace to check.
+	 * @param callable $callback           The callback to execute if the namespace should be loaded.
+	 * @param string   $rest_route         (Optional) The REST route to check against.
+	 * @param string   $callback_filter_id (Internal) Used to prevent recursive filter registration.
+	 *
+	 * @return void
+	 *
+	 * @see      self::lazy_load_namespace()
+	 * @internal Do not call this function directly. Backward compatibility is not guaranteed.
+	 */
+	public function attach_lazy_loaded_namespace( string $route_namespace, callable $callback, string $rest_route = '', string $callback_filter_id = '' ) {
+		if ( '' === $rest_route ) {
+			$rest_route = $GLOBALS['wp']->query_vars['rest_route'] ?? '';
+		}
+
+		if ( '' !== $rest_route ) {
+			$rest_route      = trailingslashit( ltrim( $rest_route, '/' ) );
+			$route_namespace = trailingslashit( $route_namespace );
+			if ( '/' === $rest_route || str_starts_with( $rest_route, $route_namespace ) ) {
+				// Load all namespaces for root requests (/wp-json/) to maintain API discovery functionality.
+				if ( '' !== $callback_filter_id ) {
+					// Remove the current filter prior to the callback, to prevent recursive callback issues.
+					// This is crucial for APIs like wc-analytics that may callback to their own namespace when loading.
+					remove_filter( 'rest_pre_dispatch', $callback_filter_id, 0 );
+				}
+
+				call_user_func( $callback );
+
+				return;
+			}
+		}
+
+		// Register a filter to check again on rest_pre_dispatch for dynamic loading.
+		if ( '' === $callback_filter_id ) {
+			$callback_filter    = function ( $filter_result, $server, $request ) use ( $route_namespace, $callback, &$callback_filter_id ) {
+				if ( is_callable( array( $request, 'get_route' ) ) ) {
+					$this->attach_lazy_loaded_namespace( $route_namespace, $callback, $request->get_route(), $callback_filter_id );
+				}
+
+				return $filter_result;
+			};
+			$callback_filter_id = _wp_filter_build_unique_id( 'rest_pre_dispatch', $callback_filter, 0 );
+			// This runs on priority 0 so that the namespace is loaded before `rest_handle_options_request()` is run (priority 10).
+			add_filter( 'rest_pre_dispatch', $callback_filter, 0, 3 );
```

Comment:
My understanding is that:

- We are passed a namespace and callback
- We create an anon function which calls `attach_lazy_loaded_namespace` with the namespace and callback
- `attach_lazy_loaded_namespace` calls the callback if the route matches the namespace, then unhooks itself

I *think* this is ok. My concern is the `recursive callback issue` mentioned in the comment above. 

Did you consider a registration type pattern whereby we hold on to the namespaces and callbacks and do a single filter on `rest_pre_dispatch` which iterates over registered callbacks and only calls those appropriate? This would mean recursion is not a problem at all.